### ISSUE TITLE
Dev dependencies are missing

### ISSUE THREAD

**Issue by acmcarther** [acmcarther/cargo-raze#7]
TL;DR: Need to extend Cargo.lock with dev dependencies -- currently examples and tests may or may not work.

It turns out that Cargo does not lock dev dependencies for declared dependencies, presumably because they are not required to compile the root crate. This is fine for the typical cargo case, but it prevents users from building examples or tests of linked crates.

It would be nice to be able to generate`rust_test` and rust_binary` for linked crates, but to do that, we'll need these dev dependencies.

More specifically -- `cargo::ops::resolve_ws_precisely` does not include dev deps. Fixing this (with custom code) is necessary, but not sufficient, to enable specifying these dependencies, because `cargo-vendor` needs to vendor them as well.

**Comment by acmcarther** [comment]
I did some research into this in a separate branch, and I found that resolving dev-dependencies as global dependencies (basically promoting them from dev to Normal) immediately creates cycles in the dependency graph.

I'm unsure if this is limited to tightly coupled dependencies, but one example is winapi 0.2.8. It creates a swarm of microdependencies that it itself depends on as dev-dependencies.

You could also trivially craft a circumstance like this -- where two testing libraries use each other for their own tests. This would reduce down to a cycle (or until one of them stops depending on the other).

From what I can tell, there are only two logical approaches to solve this:

a) Don't resolve dev dependencies. This is "easier".
b) Resolve dev dependencies in "micro envirionments". Basically, vendor dev dependencies into a separate bubble on a per-crate basis.

Option (b), would make each crate essentially the king of its own kingdom, sharing the global dependencies where appropriate, and leveraging the dev dependencies for tests. In the above example -- two test libraries using each other for tests, the hypothetical rust_test(...) invocations would not be cyclic, as they would depend on locally versioned and vendored dependencies.

**Comment by acmcarther** [comment]
Repo has moved to google/cargo-raze.

I'm not going to reopen this just yet over there, as this isn't going to get much attention, realistically. It would require a custom dependency resolution algorithm to make any real headway.
